### PR TITLE
Merge AuthZ (Tyk, Keycloak, Vault) changes to develop

### PR DESCRIPTION
**Pull request by brouillette**
Let's merge the new AuthZ stuff:
Tyk, Keycloak, and Vault integration
and their corresponding setup scripts

### CODE REVIEW

**Review by fcoralsasso — APPROVED** [review]
lgtm

File: Makefile [hunk 1]
```diff
@@ -317,6 +317,33 @@ clean-volumes:
 compose:
 	$(foreach MODULE, $(CANDIG_MODULES), $(MAKE) compose-$(MODULE);)
 
+#-- Temp --
+compose-authz:
+	docker-compose -f $(DIR)/lib/keycloak/docker-compose.yml up -d 2>&1
```
Comment (shaikh-rashid) [inline Makefile]: Would there be a need to start these services individually? If not, it is recommended to combine these under $(DIR)/lib/auth/docker-compose.yml

**Review by shaikh-rashid — CHANGES_REQUESTED** [review]
Please see change requests before closing PR

File: .gitmodules [hunk 2]
```diff
@@ -28,3 +28,6 @@
 [submodule "lib/chord-metadata/chord_metadata_service"]
 	path = lib/chord-metadata/chord_metadata_service
 	url = https://github.com/c3g/chord_metadata_service.git
+[submodule "lib/opa/opa_service"]
```
Comment (shaikh-rashid) [inline .gitmodules:31]: please consider consolidating under lib/auth/ if possible

File: Makefile [hunk 3]
```diff
@@ -317,6 +317,33 @@ clean-volumes:
...
+	docker-compose -f $(DIR)/lib/vault/docker-compose.yml up -d 2>&1
+	docker-compose -f $(DIR)/lib/tyk/docker-compose.yml up -d 2>&1
+compose-authz-down:
+	docker-compose -f $(DIR)/lib/keycloak/docker-compose.yml down
+	docker-compose -f $(DIR)/lib/vault/docker-compose.yml down
+	docker-compose -f $(DIR)/lib/tyk/docker-compose.yml down
+compose-authz-clean: compose-authz-down \
+	# needs sudo to run;
```
Comment (shaikh-rashid) [inline Makefile:326]: will need to review scripts, but sudo access will not be available in most deployment scenarios

File: Makefile [hunk 4]
```diff
@@ -317,6 +317,33 @@ clean-volumes:
...
+	./etc/setup/scripts/sudo_check.sh
+
+	# clean keycloak
+	sudo rm -r $(DIR)/lib/keycloak/volumes/* & 2>&1
```
Comment (shaikh-rashid) [inline Makefile]: use $(DIR)/tmp/data/auth instead

File: Makefile [hunk 5]
```diff
@@ -449,7 +476,8 @@ init-swarm: swarm-init swarm-networks swarm-configs swarm-secrets
 kubernetes:
 	$(DIR)/bin/kompose --file $(DIR)/lib/kubernetes/docker-compose.yml \
 		$(foreach MODULE, $(CANDIG_MODULES), --file $(DIR)/lib/$(MODULE)/docker-compose.yml) \
-		up
+		convert
```
Comment (shaikh-rashid) [inline Makefile]: please evaluate whether or not convert is required, we shouldn't need to generate kubernetes manifests

File: etc/env/example.env [hunk 6]
```diff
@@ -2,13 +2,20 @@
 # - - -
 
 # site options
-CANDIG_MODULES=weavescope portainer consul traefik logging monitoring minio mc drs-server htsget-server igv-js wes-server chord-metadata datasets cnv_service rnaget candig_server federation-service authorization jupyter
+CANDIG_MODULES=tyk keycloak vault weavescope portainer consul traefik logging monitoring minio mc drs-server htsget-server igv-js wes-server chord-metadata datasets cnv_service rnaget candig_server federation-service authorization
```
Comment (shaikh-rashid) [inline etc/env/example.env]: like logging and monitoring, auth is comprised of a collection of services. As such, it might make it easier to keep them under one library/docker-compose stack, such as in lib/auth.

File: etc/env/example.env [hunk 7]
```diff
@@ -2,13 +2,20 @@
...
+
+# candig server
+CANDIG_SERVER_VERSION=1.3.0
+CANDIG_SERVER_HOST=0.0.0.0
+CANDIG_SERVER_PORT=3001
 
 # miniconda venv
 # options are [linux, darwin]
-VENV_OS=linux
+VENV_OS=darwin
```
Comment (shaikh-rashid) [inline etc/env/example.env]: linux should be the default as it is more likely to be the default deploy environment than OS X

File: etc/env/example.env [hunk 8]
```diff
@@ -197,3 +199,47 @@ DATASETS_SERVICE_PORT=8880
 AUTHORIZATION_SERVICE_VERSION=v0.0.1-alpha
 AUTHORIZATION_SERVICE_HOST=0.0.0.0
 AUTHORIZATION_SERVICE_PORT=7000
+
+
+# keycloak service
+CONTAINER_NAME_CANDIG_AUTH=keycloak
+KC_ADMIN_USER=kcadmin
+KC_ADMIN_PW=admin
```
Comment (shaikh-rashid) [inline etc/env/example.env]: embedding secrets in global variables is not a recommended practice

File: etc/env/example.env [hunk 9]
```diff
@@ -197,3 +199,47 @@ DATASETS_SERVICE_PORT=8880
...
+KC_TEST_USER=bob
+KC_TEST_PW=bobspass123
```
Comment (shaikh-rashid) [inline etc/env/example.env]: embedding secrets in global variables is not a recommended practice

File: etc/setup/scripts/setup.sh [hunk 10]
```diff
@@ -0,0 +1,32 @@
+#! /usr/bin/env bash
```
Comment (shaikh-rashid) [inline etc/setup/scripts/setup.sh:1]: use set -e to ensure that script does not continue on fail

File: etc/setup/scripts/setup.sh [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+${PWD}/etc/setup/scripts/subtasks/vault_setup.sh
+
+# TODO: OPA
+# echo
+# echo "Setting up OPA;"
+#${PWD}/etc/setup/scripts/subtasks/opa_setup.sh
+
+echo
+echo "-- AuthZ Setup Done! --"
+echo
```
Comment (shaikh-rashid) [inline etc/setup/scripts/setup.sh:34]: looks fine, but would recommend moving scripts specific to a service in lib/ with the service

File: etc/setup/scripts/subtasks/keycloak_setup.sh [hunk 12]
```diff
@@ -0,0 +1,226 @@
+#! /usr/bin/env bash
+# This script will set up a full keycloak environment on your local CanDIGv2 cluster
```
Comment (shaikh-rashid) [inline etc/setup/scripts/subtasks/keycloak_setup.sh:4]: use set -e to ensure that script does not continue on fail

File: etc/setup/scripts/subtasks/keycloak_setup.sh [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+
+echo ">> Adding user .."
+add_user
+echo ">> .. added..."
+
+echo 
+echo ">> .. waiting for keycloak to restart..."
+while !  docker logs --tail 5  ${CONTAINER_NAME_CANDIG_AUTH} | grep "Admin console listening on http://127.0.0.1:9990" ; do sleep 1 ; done
+docker exec ${CONTAINER_NAME_CANDIG_AUTH}  rm /opt/jboss/keycloak/standalone/configuration/keycloak-add-user.json 2> /dev/null
+echo ">> .. ready..."
```
Comment (shaikh-rashid) [inline etc/setup/scripts/subtasks/keycloak_setup.sh:216]: looks fine, but would recommend moving scripts specific to a service in lib/ with the service

File: etc/setup/scripts/subtasks/opa_setup.sh [hunk 14]
```diff
@@ -0,0 +1,2 @@
+#! /usr/bin/env bash
```
Comment (shaikh-rashid) [inline etc/setup/scripts/subtasks/opa_setup.sh:1]: there is no code for opa...

File: etc/setup/scripts/subtasks/tyk_setup.sh [hunk 15]
```diff
@@ -0,0 +1,62 @@
+#! /usr/bin/env bash
+# This script will set up a full tyk environment on your local CanDIGv2 cluster
```
Comment (shaikh-rashid) [inline etc/setup/scripts/subtasks/tyk_setup.sh:4]: use set -e to ensure that script does not continue on fail

File: etc/setup/scripts/subtasks/tyk_setup.sh [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
+# ** Be sure to invoke this from the Makefile at the project's root directory [CanDigV2] **
+
+# Load Tyk template (.tpl) files, populate them 
+# with project .env variables, and then spit 
+# them out to ./lib/tyk/volumes/*
+
+mkdir -p ${PWD}/lib/tyk/volumes
```
Comment (shaikh-rashid) [inline etc/setup/scripts/subtasks/tyk_setup.sh]: use $(DIR)/tmp/data/auth instead

File: etc/setup/scripts/subtasks/tyk_setup.sh [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+
+# api_auth.json
+echo "Working on api_auth.json .."
+envsubst < ${PWD}/etc/setup/templates/configs/tyk/confs/api_auth.json.tpl > ${PWD}/lib/tyk/volumes/api_auth.json
```
Comment (shaikh-rashid) [inline etc/setup/scripts/subtasks/tyk_setup.sh]: these templates can be embedded in lib/ or under etc/tpl to generalize templates

File: etc/setup/scripts/subtasks/vault_setup.sh [hunk 18]
```diff
@@ -0,0 +1,171 @@
+#! /usr/bin/env bash
+# This script will set up a full vault environment on your local CanDIGv2 cluster
```
Comment (shaikh-rashid) [inline etc/setup/scripts/subtasks/vault_setup.sh:4]: use set -e to ensure that script does not continue on fail

File: etc/setup/scripts/subtasks/vault_setup.sh [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+# https://www.vaultproject.io/api-docs/secret/identity/entity#batch-delete-entities
+
+
+# vault-config.json
+echo "Working on vault-config.json .."
+mkdir -p ${PWD}/lib/vault/config
+envsubst < ${PWD}/etc/setup/templates/configs/vault/vault-config.json.tpl > ${PWD}/lib/vault/config/vault-config.json
+
+# boot container
+docker-compose -f ${PWD}/lib/vault/docker-compose.yml up -d
```
Comment (shaikh-rashid) [inline etc/setup/scripts/subtasks/vault_setup.sh]: call Makefile instead of running docker commands directly

File: Makefile [hunk 20]
```diff
@@ -317,6 +317,33 @@ clean-volumes:
 compose:
 	$(foreach MODULE, $(CANDIG_MODULES), $(MAKE) compose-$(MODULE);)
 
+#-- Temp --
+compose-authz:
+	docker-compose -f $(DIR)/lib/keycloak/docker-compose.yml up -d 2>&1
+	docker-compose -f $(DIR)/lib/vault/docker-compose.yml up -d 2>&1
+	docker-compose -f $(DIR)/lib/tyk/docker-compose.yml up -d 2>&1
```
Comment (shaikh-rashid) [inline Makefile]: use existing make compose-* service-* instead of creating a custom method

File: Makefile [hunk 21]
```diff
@@ -317,6 +317,33 @@ clean-volumes:
...
+compose-authz-down:
+	docker-compose -f $(DIR)/lib/keycloak/docker-compose.yml down
```
Comment (shaikh-rashid) [inline Makefile]: use existing make clean-* instead of creating a custom method

File: etc/setup/scripts/subtasks/keycloak_setup.sh [hunk 22]
```diff
@@ -0,0 +1,226 @@
...
+
+# secrets.env
+echo "Working on secrets.env .."
+envsubst < ${PWD}/etc/setup/templates/configs/keycloak/configuration/secrets.env.tpl > ${PWD}/lib/keycloak/volumes/secrets.env
+#temp
+#touch ${PWD}/lib/keycloak/volumes/secrets.env
+
+# echo 
+mkdir ${PWD}/lib/keycloak/volumes/keycloak-db
+chmod 777 ${PWD}/lib/keycloak/volumes/keycloak-db
```
Comment (shaikh-rashid) [inline etc/setup/scripts/subtasks/keycloak_setup.sh]: chmod 755 recommended over giving everyone access to the keycloak db

File: etc/setup/scripts/sudo_check.sh [hunk 23]
```diff
@@ -0,0 +1,8 @@
+#! /usr/bin/env bash
```
Comment (shaikh-rashid) [inline etc/setup/scripts/sudo_check.sh:1]: you should not need sudo, everything should be possible in userspace, including running docker commands

File: etc/setup/templates/configs/keycloak/configuration/application-roles.properties [hunk 24]
```diff
@@ -0,0 +1,23 @@
+#
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/keycloak/configuration/application-roles.properties]: there doesn't seem to be a use for this file...

File: etc/setup/templates/configs/keycloak/configuration/application-users.properties [hunk 25]
```diff
@@ -0,0 +1,25 @@
+#
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/keycloak/configuration/application-users.properties:1]: everything is commented, is this file needed then?

File: etc/setup/templates/configs/keycloak/configuration/logging.properties [hunk 26]
```diff
@@ -0,0 +1,70 @@
+#
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/keycloak/configuration/logging.properties:1]: this should be in lib/keycloak

File: etc/setup/templates/configs/keycloak/configuration/mgmt-groups.properties [hunk 27]
```diff
@@ -0,0 +1,22 @@
+#
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/keycloak/configuration/mgmt-groups.properties]: this should be in lib/keycloak, also doesn't seem to have a use

File: etc/setup/templates/configs/keycloak/configuration/mgmt-users.properties [hunk 28]
```diff
@@ -0,0 +1,27 @@
+#
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/keycloak/configuration/mgmt-users.properties:1]: please see previous comments regarding consolidating templates

File: etc/setup/templates/configs/keycloak/configuration/secrets.env.tpl [hunk 29]
```diff
@@ -0,0 +1,3 @@
+KEYCLOAK_USER=${KC_ADMIN_USER}
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/keycloak/configuration/secrets.env.tpl:1]: please see previous comments regarding consolidating templates

File: etc/setup/templates/configs/keycloak/configuration/standalone-ha.xml [hunk 30]
```diff
@@ -0,0 +1,636 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/keycloak/configuration/standalone-ha.xml:1]: please see previous comments regarding consolidating templates

File: etc/setup/templates/configs/keycloak/configuration/standalone.xml [hunk 31]
```diff
@@ -0,0 +1,573 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/keycloak/configuration/standalone.xml:1]: please see previous comments regarding consolidating templates

File: etc/setup/templates/configs/tyk/confs/api_auth.json.tpl [hunk 32]
```diff
@@ -0,0 +1,77 @@
+{
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/tyk/confs/api_auth.json.tpl:1]: please see previous comments regarding consolidating templates

File: etc/setup/templates/configs/tyk/confs/api_candig.json.tpl [hunk 33]
```diff
@@ -0,0 +1,159 @@
+{
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/tyk/confs/api_candig.json.tpl:1]: please see previous comments regarding consolidating templates

File: etc/setup/templates/configs/tyk/confs/authMiddleware.js [hunk 34]
```diff
@@ -0,0 +1,183 @@
+// ---- Authentication middleware -----
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/tyk/confs/authMiddleware.js:1]: this should be with lib/tyk

File: etc/setup/templates/configs/tyk/confs/key_request.json.tpl [hunk 35]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/tyk/confs/key_request.json.tpl:1]: this should be with lib/tyk

File: etc/setup/templates/configs/tyk/confs/oidcDistributedClaimsConduitMiddleware.js [hunk 36]
```diff
@@ -0,0 +1,140 @@
+// ---- OIDC Distributed Claims Conduit middleware -----
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/tyk/confs/oidcDistributedClaimsConduitMiddleware.js:1]: this should be with lib/tyk

File: etc/setup/templates/configs/tyk/confs/policies.json.tpl [hunk 37]
```diff
@@ -0,0 +1,39 @@
+{
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/tyk/confs/policies.json.tpl:1]: please see previous comments regarding consolidating templates

File: etc/setup/templates/configs/tyk/confs/tyk.conf.tpl [hunk 38]
```diff
@@ -0,0 +1,158 @@
+{
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/tyk/confs/tyk.conf.tpl:1]: please see previous comments regarding consolidating templates

File: etc/setup/templates/configs/tyk/confs/tyk_analytics.conf.tpl [hunk 39]
```diff
@@ -0,0 +1,92 @@
+{
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/tyk/confs/tyk_analytics.conf.tpl:1]: please see previous comments regarding consolidating templates

File: etc/setup/templates/configs/tyk/confs/virtualLogin.js [hunk 40]
```diff
@@ -0,0 +1,129 @@
+// Virtual authentication login endpoint for API Gateway
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/tyk/confs/virtualLogin.js:1]: this should be in lib/tyk

File: etc/env/example.env (hunk 9 above)
Comment (brouillette) [reply etc/env/example.env]: this is the example env, though...

File: etc/setup/templates/configs/tyk/confs/virtualLogout.js [hunk 41]
```diff
@@ -0,0 +1,39 @@
+// Virtual authentication logout endpoint for API Gateway
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/tyk/confs/virtualLogout.js:1]: this should be in lib/tyk

File: etc/setup/templates/configs/tyk/confs/virtualToken.js [hunk 42]
```diff
@@ -0,0 +1,77 @@
+// Virtual token endpoint for API Gateway
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/tyk/confs/virtualToken.js:1]: this should be in lib/tyk

File: etc/setup/templates/configs/vault/vault-config.json.tpl [hunk 43]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (shaikh-rashid) [inline etc/setup/templates/configs/vault/vault-config.json.tpl:1]: please see previous comments regarding consolidating templates

File: lib/candig_server/docker-compose.yml [hunk 44]
```diff
@@ -2,15 +2,16 @@ version: '3.7'
 
 services:
   candig-server:
+    container_name: candig-server
```
Comment (shaikh-rashid) [inline lib/candig_server/docker-compose.yml]: container name is generated with Makefile deployment, based on service name

File: lib/candig_server/docker-compose.yml [hunk 45]
```diff
@@ -2,15 +2,16 @@ version: '3.7'
...
     build:
-      context: $PWD/lib/candig_server
+      context: $PWD   #/lib/candig_server
```
Comment (shaikh-rashid) [inline lib/candig_server/docker-compose.yml]: build context should not need to be changed

File: lib/candig_server/docker-compose.yml [hunk 46]
```diff
@@ -2,15 +2,16 @@ version: '3.7'
...
       args:
         venv_python: '3.6'
-    image: ${DOCKER_REGISTRY}/candig_server:${CANDIG_SERVER_VERSION}
+    image: candig/candig_server:1.3.0
```
Comment (shaikh-rashid) [inline lib/candig_server/docker-compose.yml]: these should not be hard coded

File: lib/candig_server/docker-compose.yml [hunk 47]
```diff
@@ -2,15 +2,16 @@ version: '3.7'
...
     networks:
-      - ${DOCKER_NET}
+      - bridge-net
```
Comment (shaikh-rashid) [inline lib/candig_server/docker-compose.yml]: avoid hard coding

File: lib/candig_server/docker-compose.yml [hunk 48]
```diff
@@ -2,15 +2,16 @@ version: '3.7'
...
     ports:
-      - "${CANDIG_SERVER_PORT}:3000"
+      - "3001:3000"
```
Comment (shaikh-rashid) [inline lib/candig_server/docker-compose.yml]: avoid hard coding

File: lib/candig_server/docker-compose.yml [hunk 49]
```diff
@@ -22,12 +23,16 @@ services:
         window: 120s
       labels:
         - traefik.enable=true
-        - traefik.port=${CANDIG_SERVER_PORT}
-        - traefik.docker.network=${DOCKER_NET}
-        - traefik.frontend.rule=Host:candig-server.${CANDIG_DOMAIN}
+        - traefik.port=3001
```
Comment (shaikh-rashid) [inline lib/candig_server/docker-compose.yml]: avoid hard coding

File: lib/candig_server/docker-compose.yml [hunk 50]
```diff
@@ -22,12 +23,16 @@ services:
...
+        - traefik.docker.network=docker-compose-test-network
+        - traefik.frontend.rule=Host:candig-server.localhost
     logging:
       driver: "fluentd"
       options:
         tag: docker.{{.Name}}
         fluentd-async-connect: "true"
     command: ["--host", "0.0.0.0", "--port", "3000"]
+
+networks:
```
Comment (shaikh-rashid) [inline lib/candig_server/docker-compose.yml]: this is not needed, i assume you were doing some testing and did not clean up these changes

File: lib/keycloak/docker-compose.yml [hunk 51]
```diff
@@ -0,0 +1,31 @@
+version: '3.7'
+
+networks:
```
Comment (shaikh-rashid) [inline lib/keycloak/docker-compose.yml]: networks are already defined in lib/compose/docker-compose.yml

File: lib/candig_server/docker-compose.yml [hunk 52]
```diff
@@ -22,12 +23,16 @@ services:
...
+  bridge-net:
+    external: true
```
Comment (shaikh-rashid) [inline lib/candig_server/docker-compose.yml]: this docker-compose does not conform to specs in lib/templates

File: lib/keycloak/docker-compose.yml [hunk 53]
```diff
@@ -0,0 +1,31 @@
...
+      #- ${PWD}/lib/keycloak/volumes/secrets.env
+       - ./volumes/secrets.env
+    networks:
+      - ${DOCKER_NET}
+      # - docker-compose-test-network
+    volumes:
+      #- ${PWD}/lib/keycloak/volumes/keycloak-db:/opt/jboss/keycloak/standalone/data
+      #- ${PWD}/lib/keycloak/volumes/standalone-ha.xml:/opt/jboss/keycloak/standalone/configuration/standalone-ha.xml
+      - ./volumes/keycloak-db:/opt/jboss/keycloak/standalone/data
+      - ./volumes/standalone-ha.xml:/opt/jboss/keycloak/standalone/configuration/standalone-ha.xml
```
Comment (shaikh-rashid) [inline lib/keycloak/docker-compose.yml]: this docker-compose does not conform to specs in lib/templates

File: lib/opa/docker-compose.yml [hunk 54]
```diff
@@ -0,0 +1,37 @@
...
+      driver: "fluentd"
+      options:
+        tag: docker.{{.Name}}
+        fluentd-async-connect: "true"
+    command: ["run", "--server", "/data/metadata-service.json", "/data/metadata-service.rego"]
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost:8181/health"]
+      interval: 30s
+      timeout: 20s
+      retries: 3
```
Comment (shaikh-rashid) [inline lib/opa/docker-compose.yml]: opa docker-compose.yml looks good, but see previous comments about consolidating authN/Z services

File: lib/tyk/docker-compose.yml [hunk 55]
```diff
@@ -0,0 +1,46 @@
+version: '3.7'
+
+services:
+  tyk:
+    container_name: tyk
+    #build:
+    #  context: . # ${PWD}/lib/tyk
+    image: tykio/tyk-gateway:v2.9.3.1
```
Comment (shaikh-rashid) [inline lib/tyk/docker-compose.yml]: pin versions in .env, avoid hard coding

File: lib/tyk/docker-compose.yml [hunk 56]
```diff
@@ -0,0 +1,46 @@
+version: '3.7'
+
+services:
+  tyk:
+    container_name: tyk
```
Comment (shaikh-rashid) [inline lib/tyk/docker-compose.yml]: container name should not be needed

File: lib/tyk/docker-compose.yml [hunk 57]
```diff
@@ -0,0 +1,46 @@
...
+    container_name: tyk-redis
+    image: redis:4.0.14-alpine
+    networks:
+     - ${DOCKER_NET}
+    volumes:
+      - ./volumes/redis-data:/data
+
+networks:
+  bridge-net:
+    external: true
```
Comment (shaikh-rashid) [inline lib/tyk/docker-compose.yml]: this docker-compose does not conform to specs in lib/templates

File: lib/vault/Dockerfile [hunk 58]
```diff
@@ -0,0 +1,45 @@
+# From https://www.bogotobogo.com/DevOps/Docker/Docker-Vault-Consul.php
+# base image
+FROM alpine:3.7
+
+# set vault version
+ENV VAULT_VERSION 1.5.0
```
Comment (shaikh-rashid) [inline lib/vault/Dockerfile]: this should be a runtime arg

File: lib/vault/Dockerfile [hunk 59]
```diff
@@ -0,0 +1,45 @@
...
+
+# maintainer
+LABEL maintainer="brennan.brouillette@c3g.ca"
+
+# create a new directory
+RUN mkdir /vault
+RUN mkdir /vault/config
```
Comment (shaikh-rashid) [inline lib/vault/Dockerfile]: this can be mkdir -p /vault/config

File: lib/vault/docker-compose.yml [hunk 60]
```diff
@@ -0,0 +1,32 @@
...
+    networks:
+      - ${DOCKER_NET}
+      # - docker-compose-test-network
+
+networks:
+  bridge-net:
+    external: true
+# networks:
+#   docker-compose-test-network:
+#     external: true
```
Comment (shaikh-rashid) [inline lib/vault/docker-compose.yml]: this docker-compose does not conform to specs in lib/templates

File: etc/setup/scripts/subtasks/opa_setup.sh (hunk 14 above)
Comment (brouillette) [reply etc/setup/scripts/subtasks/opa_setup.sh:1]: right, this shouldn't be part of the pr

File: lib/tyk/docker-compose.yml (hunk 56 above)
Comment (brouillette) [reply lib/tyk/docker-compose.yml]: how would services communicate without a name to resolve to?

File: lib/tyk/docker-compose.yml (hunk 57 above)
Comment (brouillette) [reply lib/tyk/docker-compose.yml]: without this network tag, i am faced with errors stating the network doesn't exist.
any ideas?

File: lib/opa/docker-compose.yml (hunk 54 above)
Comment (brouillette) [reply lib/opa/docker-compose.yml]: this should not be part of this PR

File: Makefile (hunk 5 above)
Comment (brouillette) [reply Makefile]: this is a mistake - should not be part of this PR

File: Makefile (hunk 3 above)
Comment (brouillette) [reply Makefile:326]: I could not find another way to remove the necessary directories. Any suggestions?

File: etc/env/example.env (hunk 6 above)
Comment (brouillette) [reply etc/env/example.env]: good point, thank you

File: Makefile (hunk 4 above)
Comment (brouillette) [reply Makefile]: will begin using authz module naming in a parallel dev branch :+1:

File: etc/setup/scripts/setup.sh (hunk 11 above)
Comment (brouillette) [reply etc/setup/scripts/setup.sh:34]: will look into this in a parallel dev branch

File: etc/setup/scripts/subtasks/tyk_setup.sh (hunk 17 above)
Comment (brouillette) [reply etc/setup/scripts/subtasks/tyk_setup.sh]: will consider this in another branch

File: etc/setup/templates/configs/tyk/confs/authMiddleware.js (hunk 34 above)
Comment (brouillette) [reply etc/setup/templates/configs/tyk/confs/authMiddleware.js:1]: will consider this in the next dev branch

**Review by brouillette — COMMENTED** [review]
Many changes were pushed, and others will be made on a separate dev branch that extends this work which will soon be merged again to `develop`
